ddtrace/debugging/_signal/context.py

```diff
@@ -0,0 +1,95 @@
+from collections import ChainMap
+import typing as t
+
+from ddtrace.debugging._probe.model import ProbeConditionMixin
+from ddtrace.debugging._probe.model import ProbeEvalTiming
+from ddtrace.debugging._probe.model import RateLimitMixin
+from ddtrace.debugging._probe.model import TimingMixin
+from ddtrace.debugging._signal.model import Signal
+from ddtrace.debugging._signal.model import SignalState
+from ddtrace.internal.compat import ExcInfoType
+from ddtrace.internal.rate_limiter import RateLimitExceeded
+
+
+class SignalContext:
+    """Debugger signal context manager for function probes.
+
+    Probes with an evaluation timing will trigger the following logic:
+
+    - If the timing is on ENTRY, the probe will trigger on entry and on exit,
+      but the condition and rate limiter will only be evaluated on entry.
+
+    - If the timing is on EXIT, the probe will trigger on exit only, and the
+      condition and rate limiter will be evaluated on exit.
+
+    - IF the timing is DEFAULT, the probe will trigger on the default timing for
+      the signal.
+
+    - If the probe has no timing, it defaults to the ENTRY timing.
+    """
+
+    def __init__(self, signal: Signal) -> None:
+        self.signal = signal
+        self._timing = ProbeEvalTiming.ENTRY
+
+        probe = signal.probe
+        if isinstance(probe, TimingMixin):
+            eval_at = probe.evaluate_at
+            self._timing = signal.__default_timing__ if eval_at is ProbeEvalTiming.DEFAULT else eval_at
+
+    def enter(self) -> None:
+        signal = self.signal
+        probe = signal.probe
+
+        if self._timing is not ProbeEvalTiming.ENTRY:
+            return
+
+        scope = ChainMap(signal.args, signal.frame.f_globals)
+        if isinstance(probe, ProbeConditionMixin) and not signal._eval_condition(scope):
+            return
```

Comment:
if this is a bug?

we can have a probe with a condition, the condition is false. it return without setting signal.state
at the exit() - we check if signal == None - go ahead and capture.

so, probe with eval==ENTRY will capture always on exit. I think we should mark this as SKIP_COND?

Reply:
`_eval_condition` sets the state https://github.com/DataDog/dd-trace-py/blob/1d33870b84922eaf89136b75e7e90a048f3fa688/ddtrace/debugging/_signal/model.py#L83 so we should be good